src/app/qgisapp.cpp

```diff
@@ -9726,7 +9726,38 @@ void QgisApp::mergeSelectedFeatures()
                          vl->dataProvider()->pkAttributeIndexes().contains( vl->fields().fieldOriginIndex( i ) );
 
     if ( isPrimaryKey && !isDefaultValue )
-      mergeFeatureId = val.toLongLong();
+    {
+      QgsFeatureRequest request;
+      request.setFlags( QgsFeatureRequest::Flag::NoGeometry );
+      // Handle multi pks
+      if ( vl->dataProvider()->pkAttributeIndexes().count() > 1 && vl->dataProvider()->pkAttributeIndexes().count() <= attrs.count() )
```

Comment:
I'm not saying this fix is wrong, but this looks very complex to sit at such a high level component (QgisApp). Do you think it's possible to move this logic into a virtual method in QgsVectorDataProvider? Potentially the data provider can determine this internally without having to resort to a feature request at all, since it will already have internal hashes of attribute(s) to feature ids.

Reply:
I agree: this logic should not sit in apps and should be CI tested.

I'm just saying that I did this bugfix in my volunteer time because I had a few spare hours, I don't think I have time now for a deeper refactoring.

This is a serious regression IMO, and was even backported.

Reply:
How about we just revert the original PR until a better process is available? (Or someone funds a proper fix for the original issue)

Reply:
Ping @3nids @andreasneumann  @domi4484

Reply:
@domi4484 @elpaso @nyalldawson

Of course we are not amused to pay a fix for a fix - and also having to wait until June 2023 to get this in 3.28.x again, because every fix has to skip one release (quarantine rule). I will have a hard time explaining this to my users.

But that's life - and if you think this has to be reverted, then it has to be done ...

Reply:
@elpaso I compiled your fix and could test it successfully. It also fixes another unreported error we recently observed. And I suspect it could also fix #48402.

At this point I would ask if it is possible to merge this P.R., because reverting #47822 would result in a regression in 3.28 losing the fix to #47629.

We are currently checking possibilities to follow up with a proper fix and test in the near future.

Reply:
@domi4484 I'm ok as a temporary fix. If @nyalldawson is ok we can merge (and backport).

Reply:
I won't block it -- but @domi4484  please recognise that it's a significant amount of high maintenance, non-unit-tested code being added, and your follow up improvements would be much welcomed.

Reply:
Yes I see it and I am sorry for the inconveniences and amount of work that the regression raised up.

Reply:
@domi4484 do you want me to merge this PR?

Reply:
@elpaso I have planned to work on this next week (refactor + tests). But we could already merge this so we are sure that something is there before the freeze?